localstack/services/kms/provider.py

```diff
@@ -0,0 +1,458 @@
+        region_details = KMSBackend.get()
+
+        grant = dict(create_grant_request)
+        grant[GRANT_ID] = long_uid()
+        grant[GRANT_TOKENS] = [long_uid()]
+        if NAME not in grant:
+            grant[NAME] = ""
+        grant[CREATION_DATE] = time.time()
+
+        region_details.grants[grant[GRANT_ID]] = grant
+        return CreateGrantResponse(GrantId=grant[GRANT_ID], GrantToken=grant[GRANT_TOKENS][0])
+
+    @handler("ListGrants", expand=False)
+    def list_grants(
+        self, context: RequestContext, list_grants_request: ListGrantsRequest
+    ) -> ListGrantsResponse:
+        region_details = KMSBackend.get()
+        verify_key_exists(list_grants_request[KEY_ID])
+
+        limit = list_grants_request.get("Limit", 50)
+
+        if "Marker" in list_grants_request:
+            filtered = region_details.markers.get(list_grants_request["Marker"], [])
+        else:
+            filtered = [
+                grant
+                for grant in region_details.grants.values()
+                if grant[KEY_ID] == list_grants_request[KEY_ID]
+                and filter_grant_id(grant, list_grants_request)
+                and filter_grantee_principal(grant, list_grants_request)
+            ]
+
+        # filter out attributes
+        filtered = [remove_attributes(dict(grant), ["GrantTokens"]) for grant in filtered]
+
+        if len(filtered) <= limit:
+            return ListGrantsResponse(Grants=filtered, Truncated=False)
+
+        in_limit = filtered[:limit]
+        out_limit = filtered[limit:]
+
+        marker_id = long_uid()
+        region_details.markers[marker_id] = out_limit
+
+        return ListGrantsResponse(Grants=in_limit, Truncated=True, NextMarker=marker_id)
+
+    def revoke_grant(
+        self, context: RequestContext, key_id: KeyIdType, grant_id: GrantIdType
+    ) -> None:
+        grants = KMSBackend.get().grants
+        assert grants[grant_id][KEY_ID] == key_id
+        grants.pop(grant_id)
+
+    def retire_grant(
+        self,
+        context: RequestContext,
+        grant_token: GrantTokenType = None,
+        key_id: KeyIdType = None,
+        grant_id: GrantIdType = None,
+    ) -> None:
+        region_details = KMSBackend.get()
+        grants = region_details.grants
+
+        if grant_id and grants[grant_id][KEY_ID] == key_id:
+            grants.pop(grant_id)
+        elif grant_token:
+            region_details.grants = {
+                grant_id: grant
+                for grant_id, grant in grants.items()
+                if grant_token not in grant[GRANT_TOKENS]
+            }
+        else:
+            raise InvalidGrantTokenException("Grant token OR (grant ID, key ID) must be specified")
+
+    def list_retirable_grants(
+        self,
+        context: RequestContext,
+        retiring_principal: PrincipalIdType,
+        limit: LimitType = None,
+        marker: MarkerType = None,
+    ) -> ListGrantsResponse:
+        region_details = KMSBackend.get()
```

Comment:
As with `list_grants`, the previous implementation checked if the (required) parameter (in this case `retiring_principal`) was set (see comment above).